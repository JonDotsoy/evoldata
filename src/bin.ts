import * as fs from "fs/promises";
import { mainHelp, mainParser } from "./cmds/main-parser";
import { parse } from ".";
import * as YAML from "yaml";
import * as JSON5 from "json5";
import { format } from "prettier";
import { FormatTypes } from "./format-types/format-types";

const main = async () => {
  const { scriptFile, format, showHelp } = mainParser(process.argv.slice(2));

  if (!scriptFile || showHelp) return console.log(mainHelp(process.argv[1]));

  const out = parse(await fs.readFile(scriptFile, "utf-8"));

  if (format === FormatTypes.JSON)
    return console.log(JSON.stringify(out, null, 2));
  if (format === FormatTypes.JSON5)
    return console.log(JSON5.stringify(out, null, 2));
  if (format === FormatTypes.YAML)
    return console.log(YAML.stringify(out, null, 2));
  // console.log("ðŸš€ ~ main ~ out:", out)
};

await main();
