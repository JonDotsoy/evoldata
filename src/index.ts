import { readableStreamToIterable, SplitStream } from "streamable-tools";

type Payload = string;
type Path = number | string;
type Metadata = {
  timestamp: number;
  type: "=" | "+";
  path: Path[];
  value: unknown;
};

export namespace utils {
  const isRecord = (
    value: unknown,
  ): value is Record<string | string | number, unknown> =>
    typeof value === "object" && value !== null;

  export const selectChild = (
    object: unknown,
    paths: Path[],
    initLastPath: () => any = () => undefined,
  ): unknown => {
    const objectIsRecord = isRecord(object);
    if (!objectIsRecord) return;
    if (paths.length <= 0) return object;

    const [property, ...restPaths] = paths;
    const isLastPath = restPaths.length <= 0;

    const child = isRecord(object[property])
      ? object[property]
      : isLastPath
        ? initLastPath()
        : {};
    object[property] = child;

    return selectChild(child, restPaths, initLastPath);
  };

  export const selectRecordChild = (object: unknown, paths: Path[]) => {};

  export const set = (object: unknown, paths: Path[], value: unknown) => {
    const childPath = paths.splice(0, paths.length - 1);
    const property = paths[paths.length - 1];

    const child = selectChild(object, childPath, () => ({}));

    if (isRecord(child)) {
      child[property] = value;
    }
  };

  export const add = (object: unknown, paths: Path[], value: unknown) => {
    if (paths.length < 1) return;
    const parentPaths = paths.splice(0, paths.length - 1);
    const property = paths[paths.length - 1];
    const parentSelect = selectChild(object, parentPaths, () => ({}));
    const parent: any = isRecord(parentSelect) ? parentSelect : {};
    const childIsArray = Array.isArray(parent[property]);

    if (childIsArray) {
      parent[property].push(value);
    } else {
      parent[property] = [value];
    }
  };

  export namespace path {
    export const serialize = (paths: Path[]) =>
      paths
        .map((path) =>
          `${path}`.replace(/\W/g, (c) =>
            [".", "$"].includes(c) ? `$${c.charCodeAt(0)}` : c,
          ),
        )
        .join(".");

    export const deserialize = (path: string) =>
      path
        .split(".")
        .map((part) =>
          part.replace(/\$(\d+)/, (_, c) => String.fromCharCode(Number(c))),
        )
        .map((e) => (/^\d+$/.test(e) ? Number(e) : e));
  }
}

const payloadToReadable = (payload: Payload) => {
  if (typeof payload === "string") return new TextEncoder().encode(payload);

  throw new Error("Payload must be a string");
};

const TAB_CODE = "\t".charCodeAt(0);
const NEW_LINE_CODE = "\n".charCodeAt(0);

function* transformLines(buff: Uint8Array): Generator<Metadata> {
  let index = 0;

  const findPart = (delimiter: number) => {
    const a = buff.indexOf(delimiter, index);
    const po = a === -1 ? buff.length : a;
    const part = buff.slice(index, po);
    index = po + 1;
    const chunk = new TextDecoder().decode(part);
    return chunk;
  };

  while (index < buff.length) {
    const timestampBuff = Number(findPart(TAB_CODE));
    const action = findPart(TAB_CODE);
    const path = utils.path.deserialize(findPart(TAB_CODE));
    const value = JSON.parse(findPart(NEW_LINE_CODE));

    yield {
      timestamp: timestampBuff,
      type: action as "=" | "+",
      path,
      value,
    };
  }
}

export const parse = (payload: Payload): any => {
  const obj: any = {};

  for (const a of transformLines(payloadToReadable(payload))) {
    if (a.type === "=") {
      utils.set(obj, a.path, a.value);
    }
    if (a.type === "+") {
      utils.add(obj, a.path, a.value);
    }
  }

  return obj;
};

export const stringifyLineEvent = (metadata: Metadata) => {
  const { timestamp, type, path, value } = metadata;
  return `${timestamp}\t${type}\t${utils.path.serialize(path)}\t${JSON.stringify(value)}\n`;
};

export const createEventsWritable = () => {
  let controller: ReadableStreamDefaultController<Uint8Array> | null = null;
  const readable = new ReadableStream<Uint8Array>({
    start: (ctr) => {
      controller = ctr;
    },
  });

  const makeEvent = (action: string, paths: Path[], value: unknown) => {
    const chunk = stringifyLineEvent({
      timestamp: Date.now(),
      type: action as "=" | "+",
      path: paths,
      value,
    });

    controller?.enqueue(new TextEncoder().encode(chunk));
  };

  return {
    readable,
    close: () => {
      controller?.close();
    },
    set(path: Path[], value: unknown) {
      makeEvent("=", path, value);
    },
    add(path: Path[], value: unknown) {
      makeEvent("+", path, value);
    },
  };
};

export class ParsingObjectStream extends TransformStream<Uint8Array, any> {
  #obj: { context: any } = { context: {} };

  constructor() {
    super({
      transform: async (chunk, controller) => {
        try {
          for (const metadata of transformLines(chunk)) {
            if (metadata.type === "=") {
              utils.set(this.#obj.context, metadata.path, metadata.value);
            }
            if (metadata.type === "+") {
              utils.add(this.#obj.context, metadata.path, metadata.value);
            }
            controller.enqueue(this.#obj.context);
          }
        } catch {}
      },
    });
  }

  static iterable(readable: ReadableStream<Uint8Array>) {
    return readableStreamToIterable(
      readable
        .pipeThrough(new SplitStream())
        .pipeThrough(new ParsingObjectStream()),
    );
  }

  static store(readable: ReadableStream<Uint8Array>) {
    const snapshot: { current: any } = { current: {} };
    const listeners = new Set<() => void>();

    const loop = async () => {
      for await (const snap of ParsingObjectStream.iterable(readable)) {
        snapshot.current = snap;
        listeners.forEach((listener) => listener());
      }
    };

    loop()
      .catch((err) => console.error(err))
      .finally();

    const subscribe = (listener: () => void) => {
      listeners.add(listener);
      return () => listeners.delete(listener);
    };
    const getSnapshot = () => snapshot.current;

    return { subscribe, getSnapshot };
  }
}
