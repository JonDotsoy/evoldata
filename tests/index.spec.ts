import { describe, it, expect } from "bun:test";
import { parse, utils, createEventsWritable } from "../src";

const payload = `1730041100000	=	owner.name	${JSON.stringify("John")}
1730041100000	=	owner.runOn	${JSON.stringify("Ubuntu 30.04")}
1730041200000	=	server.ip	${JSON.stringify("10.0.0.1")}
1730041200000	=	server.port	${JSON.stringify(22)}
1730041300000	=	client.state	${JSON.stringify("stable")}
1730041300000	=	client.pid	${JSON.stringify(1000)}
1730041300000	+	client.logs	${JSON.stringify("Connected to server.")}
1730041400000	=	client.delay	${JSON.stringify(100)}
1730041500000	=	client.state	${JSON.stringify("unstable")}
1730041500000	+	server.commands	${JSON.stringify("date")}
1730041600000	+	client.logs	${JSON.stringify("Sun Oct 27 12:12:48 -03 2024\n")}
`;

describe("utils", () => {
  it("set() should create nested objects and set the value at the specified path", () => {
    const obj = {};
    utils.set(obj, ["a", "b", "c"], "abc");

    expect(obj).toEqual({
      a: {
        b: {
          c: "abc",
        },
      },
    });
  });

  it("set() should create nested objects with numeric keys and set the value", () => {
    const obj = {};
    utils.set(obj, ["a", 0, "c"], "abc");

    expect(obj).toEqual({
      a: {
        0: {
          c: "abc",
        },
      },
    });
  });

  it("add() should create nested objects with numeric keys and initialize an array with the value", () => {
    const obj = {};
    utils.add(obj, ["a", 0, "c"], "abc");

    expect(obj).toEqual({
      a: {
        0: {
          c: ["abc"],
        },
      },
    });
  });

  it("set() should add a new property to an existing nested object", () => {
    const obj: any = { a: { 0: { c: { d: "abc" } } } };
    utils.set(obj, ["a", 0, "e"], "abc");

    expect(obj).toEqual({ a: { 0: { c: { d: "abc" }, e: "abc" } } });
  });

  it("add() should overwrite existing property with an array if the property is not an array", () => {
    const obj: any = { a: { 0: { c: { d: "abc" } } } };
    utils.add(obj, ["a", 0, "c"], "abc");

    expect(obj).toEqual({ a: { 0: { c: ["abc"] } } });
  });

  it("add() should initialize an array with the value if the path does not exist", () => {
    const obj: any = {};
    utils.add(obj, ["a"], "abc");

    expect(obj).toEqual({ a: ["abc"] });
  });

  it("add() should append multiple values to an array at the specified path", () => {
    const obj: any = {};
    utils.add(obj, ["a", 0, "c"], "abc");
    utils.add(obj, ["a", 0, "c"], "def");
    utils.add(obj, ["a", 0, "c"], "zzz");

    expect(obj).toEqual({
      a: {
        0: {
          c: ["abc", "def", "zzz"],
        },
      },
    });
  });

  it("at() should create nested objects with undefined values for non-existing paths", () => {
    const obj = {};
    utils.selectChild(obj, ["a", "b", "c"]);

    expect(obj).toEqual({ a: { b: { c: undefined } } });
  });
});

describe("path", () => {
  describe("serialize", () => {
    it("should serialize a path with special characters", () => {
      expect(utils.path.serialize(["a", ".", "b"])).toEqual("a.$46.b");
    });
    it("should serialize a path with multiple special characters", () => {
      expect(utils.path.serialize(["a", "$", ".", "b"])).toEqual("a.$36.$46.b");
    });
    it("should serialize a path with special characters and starting with '-'", () => {
      expect(utils.path.serialize(["-", "$", ".", "[]"])).toEqual(
        "-.$36.$46.[]",
      );
    });
    it("should serialize a path with special characters, starting with '-' and containing a number", () => {
      expect(utils.path.serialize(["-", 10, "$", ".", "[]"])).toEqual(
        "-.10.$36.$46.[]",
      );
    });
  });
  describe("desearilize", () => {
    it("should deserialize a path with special characters", () => {
      expect(utils.path.deserialize("a.$46.b")).toEqual(["a", ".", "b"]);
    });
    it("should deserialize a path with multiple special characters", () => {
      expect(utils.path.deserialize("a.$36.$46.b")).toEqual([
        "a",
        "$",
        ".",
        "b",
      ]);
    });
    it("should deserialize a path with special characters and starting with '-'", () => {
      expect(utils.path.deserialize("-.$36.$46.[]")).toEqual([
        "-",
        "$",
        ".",
        "[]",
      ]);
    });
    it("should deserialize a path with special characters, starting with '-' and containing a number", () => {
      expect(utils.path.deserialize("-.10.$36.$46.[]")).toEqual([
        "-",
        10,
        "$",
        ".",
        "[]",
      ]);
    });
  });
});

it("parse() should parse a payload string into an object", () => {
  const obj = parse(payload);

  expect(obj).toMatchSnapshot();
});

it("createEventsWritable() should create a writable stream that emits events as strings", async () => {
  const writable = createEventsWritable();

  writable.set(["name"], "Jhon");
  writable.set(["age"], 32);

  writable.close();

  const read = writable.readable.pipeThrough(new TextDecoderStream());

  const text = await Bun.readableStreamToText(read);

  const lines = text.split("\n");
  expect(lines.at(0)).toMatch('=\tname\t"Jhon"');
  expect(lines.at(1)).toMatch("=\tage\t32");
});
