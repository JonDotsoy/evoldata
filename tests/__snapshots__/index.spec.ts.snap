// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parse() should parse a payload string into an object 1`] = `
{
  "client": {
    "delay": 100,
    "logs": [
      "Connected to server.",
      
"Sun Oct 27 12:12:48 -03 2024
"
,
    ],
    "pid": 1000,
    "state": "unstable",
  },
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04",
  },
  "server": {
    "commands": [
      "date",
    ],
    "ip": "10.0.0.1",
    "port": 22,
  },
}
`;

exports[`StreamSnapParsing should parse a stream of events into an array of objects 1`] = `
{
  "owner": {
    "name": "John",
  },
}
`;

exports[`StreamSnapParsing should parse a stream of events into an array of objects 2`] = `
{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04",
  },
}
`;

exports[`StreamSnapParsing should parse a stream of events into an array of objects 3`] = `
{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04",
  },
  "server": {
    "ip": "10.0.0.1",
  },
}
`;

exports[`StreamSnapParsing should parse a stream of events into an array of objects 4`] = `
{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04",
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22,
  },
}
`;

exports[`StreamSnapParsing should parse a stream of events into an array of objects 5`] = `
{
  "client": {
    "state": "stable",
  },
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04",
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22,
  },
}
`;

exports[`StreamSnapParsing should parse a stream of events into an array of objects 6`] = `
{
  "client": {
    "pid": 1000,
    "state": "stable",
  },
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04",
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22,
  },
}
`;

exports[`StreamSnapParsing should parse a stream of events into an array of objects 7`] = `
{
  "client": {
    "logs": [
      "Connected to server.",
    ],
    "pid": 1000,
    "state": "stable",
  },
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04",
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22,
  },
}
`;

exports[`StreamSnapParsing should parse a stream of events into an array of objects 8`] = `
{
  "client": {
    "delay": 100,
    "logs": [
      "Connected to server.",
    ],
    "pid": 1000,
    "state": "stable",
  },
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04",
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22,
  },
}
`;

exports[`StreamSnapParsing should parse a stream of events into an array of objects 9`] = `
{
  "client": {
    "delay": 100,
    "logs": [
      "Connected to server.",
    ],
    "pid": 1000,
    "state": "unstable",
  },
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04",
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22,
  },
}
`;

exports[`StreamSnapParsing should parse a stream of events into an array of objects 10`] = `
{
  "client": {
    "delay": 100,
    "logs": [
      "Connected to server.",
    ],
    "pid": 1000,
    "state": "unstable",
  },
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04",
  },
  "server": {
    "commands": [
      "date",
    ],
    "ip": "10.0.0.1",
    "port": 22,
  },
}
`;

exports[`StreamSnapParsing should parse a stream of events into an array of objects 11`] = `
{
  "client": {
    "delay": 100,
    "logs": [
      "Connected to server.",
    ],
    "pid": 1000,
    "state": "unstable",
  },
  "done": true,
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04",
  },
  "server": {
    "commands": [
      "date",
    ],
    "ip": "10.0.0.1",
    "port": 22,
  },
}
`;

exports[`ParsingObjectStream.store() should create a store that updates on each event 1`] = `
[
  [
    "{}",
  ],
  [
    
"{
  "owner": {
    "name": "John"
  }
}"
,
  ],
  [
    
"{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04"
  }
}"
,
  ],
  [
    
"{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04"
  },
  "server": {
    "ip": "10.0.0.1"
  }
}"
,
  ],
  [
    
"{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04"
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22
  }
}"
,
  ],
  [
    
"{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04"
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22
  },
  "client": {
    "state": "stable"
  }
}"
,
  ],
  [
    
"{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04"
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22
  },
  "client": {
    "state": "stable",
    "pid": 1000
  }
}"
,
  ],
  [
    
"{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04"
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22
  },
  "client": {
    "state": "stable",
    "pid": 1000,
    "logs": [
      "Connected to server."
    ]
  }
}"
,
  ],
  [
    
"{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04"
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22
  },
  "client": {
    "state": "stable",
    "pid": 1000,
    "logs": [
      "Connected to server."
    ],
    "delay": 100
  }
}"
,
  ],
  [
    
"{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04"
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22
  },
  "client": {
    "state": "unstable",
    "pid": 1000,
    "logs": [
      "Connected to server."
    ],
    "delay": 100
  }
}"
,
  ],
  [
    
"{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04"
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22,
    "commands": [
      "date"
    ]
  },
  "client": {
    "state": "unstable",
    "pid": 1000,
    "logs": [
      "Connected to server."
    ],
    "delay": 100
  }
}"
,
  ],
  [
    
"{
  "owner": {
    "name": "John",
    "runOn": "Ubuntu 30.04"
  },
  "server": {
    "ip": "10.0.0.1",
    "port": 22,
    "commands": [
      "date"
    ]
  },
  "client": {
    "state": "unstable",
    "pid": 1000,
    "logs": [
      "Connected to server."
    ],
    "delay": 100
  },
  "done": true
}"
,
  ],
]
`;
